libjolie/src/main/java/jolie/lang/parse/module/ModuleCrawler.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (C) 2020 Narongrit Unwerawattana <narongrit.kie@gmail.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301  USA
+ */
+
+package jolie.lang.parse.module;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import jolie.lang.parse.ParserException;
+import jolie.lang.parse.module.exceptions.ModuleNotFoundException;
+
+public class ModuleCrawler {
+
+	public static class CrawlerResult {
+		private final Map< URI, ModuleRecord > moduleCrawled;
+
+		private CrawlerResult() {
+			this.moduleCrawled = new HashMap<>();
+		}
+
+		private void addModuleRecord( ModuleRecord mr ) {
+			this.moduleCrawled.put( mr.source(), mr );
+		}
+
+		private boolean isRecordInResult( URI source ) {
+			return this.moduleCrawled.containsKey( source );
+		}
+
+		public Map< URI, ModuleRecord > toMap() {
+			return this.moduleCrawled;
+		}
+
+		public Map< URI, SymbolTable > symbolTables() {
+			Map< URI, SymbolTable > result = new HashMap<>();
+			for( ModuleRecord mr : this.moduleCrawled.values() ) {
+				result.put( mr.source(), mr.symbolTable() );
+			}
+			return result;
+		}
+	}
+
+	private static final Map< URI, ModuleRecord > cache = new ConcurrentHashMap<>();
+
+	private static void putToCache( ModuleRecord mc ) {
+		ModuleCrawler.cache.put( mc.source(), mc );
+	}
+
+	private static boolean inCache( URI source ) {
+		return ModuleCrawler.cache.containsKey( source );
+	}
+
+	private static ModuleRecord getRecordFromCache( URI source ) {
+		return ModuleCrawler.cache.get( source );
+	}
+
+	public static void clearCache() {
+		ModuleCrawler.cache.clear();
+	}
+
+	private final Finder finder;
+	private final ModuleParser parser;
+
+	private ModuleCrawler( ModuleCrawlerComponent component ) {
```

Comment:
Erase the `ModuleCrawlerComponent` class and just bring the `Finder` and `ParserConfiguration` objects at this level.